src/WebJobs.Script/Host/ScriptHostManager.cs

```diff
@@ -342,6 +362,24 @@ private void OnScriptFileChanged(object sender, FileSystemEventArgs e)
             }
         }
 
+        // Assume called once per ScriptHostManager
+        public void StopThenDispose()
```

Comment:
Can we scale back these changes for this fix, and make any additional synchronization changes you're proposing later?

Specifically, I think you should add a StopAndDisposeAsync method that does something simple like our existing method used for ScriptHosts. For the time being, use of this might result in object disposed exceptions in some cases, but that isn't a huge issue for now. Later, we can change the internal impl. of StopAndDisposeAsync to address those issues as well.